### ISSUE TITLE
Add $Dependencies that references all dependent classes

### ISSUE THREAD

**Issue by plasmaboyer** [frhagn/Typewriter#52]
```csharp

public class MyController : Controller {
   public MyViewModel Get() {
     //...
     return new MyViewModel();
  }
}
```

```
$Classes(*Controller) [
   $Dependencies [ 
      $Classes [ 
        //Generate dependencies for instance MyViewModel
      ]
   ]
]
```

**Comment by frhagn** [comment]
Hi @plasmaboyer.
Interesting idea to extract the dependencies from a class. I think this is difficult to implement in a reliable way, especially with CodeDom, but it might be possible. 

Maybe you could try something like this to solve it instead?
```
$Classes(*Controller)[
    $Methods[        
        $Type[interface $Name { $Properties[
            $name: $Type;]
        }]
    ]
]
```

**Comment by plasmaboyer** [comment]
Thanks @frhagn for your answer.
It works but not recursively : if $Type has a property with an non-generated type, it does not work

I'm trying to create a custom method that create a Dictionary of classes that must be generated.

**Comment by frhagn** [comment]
Hi @plasmaboyer!
I've added a $IsDefined property in Typewriter 1.1, so now you could try someting like this to solve your problem:
```
${
    using Typewriter.Extensions.Types;
        
    IEnumerable<Type> DefinedTypes(Class c) 
    {
        return c.Methods.Where(m => m.Type.Unwrap().IsDefined).Select(m => m.Type.Unwrap());
    }
}
$Classes(*Controller)[$DefinedTypes[
    interface $Name { $Properties[
        $name: $Type;]
    }]
]
```

Hope it helps!

**Comment by cwoolum** [comment]
I tried this and I'm running into an issue where if i use a model in multiple controllers, it generates the definition multiple times.and breaks... I think it needs to go one level higher. Is it possible to go up to the $Classes level?

**Comment by frhagn** [comment]
You can go up to the file level, but that doesn't really solve much in this case. Since the templates are rendered one file at a time, Typewriter doesn't know about anything outside the current source file.

My recommended solution is to use two separate tst templates with their own specific purpose; one for creating the controllers and one for the models.

**Comment by cwoolum** [comment]
I am using a separate models templates but I don't want to generate typescript definitions for every model in the namespace. It would be ideal to only generate models for the Types returned by the controllers and their referenced types. I'm currently using an attribute to determine if a Class should be generated but its annoying to have to add the attribute to every Class I intend to send to the client.

**Comment by frhagn** [comment]
I agree, that would be a nice feature, but I can't see any way to implement it at the moment. I'll try to work someting out, but I don't expect to be able to solve this anytime soon, sorry.

**Comment by plasmaboyer** [comment]
Does this refactoring could work ? (code not tested) :

```csharp
static HashSet<Type> _foundTypes = new HashSet<Type>();

Enumerable<Type> DefinedTypes(Class c) 
    {
       var values = c.Methods.Where(m => m.Type.Unwrap().IsDefined).Select(m => m.Type.Unwrap());
       var newValues = values.Where(t=> !_foundTypes.Contains(t));
       _foundTypes.UnionWith(newValues);

      return newValues;
    }
```

**Comment by cwoolum** [comment]
unfortunately, that would still produce duplicates

**Comment by agustinbcu01** [comment]
IEnumerable <Type> MethodDependencies(Method m, bool getEnums){
        return TypeAndGenericTypes(m.Type)
                .Union(m.Parameters.SelectMany(p => TypeAndGenericTypes(p.Type)))
                .Where(t => t.IsDefined && t.FullName.Contains("Model name space ") && (getEnums == t.IsEnum));
    }
    IEnumerable<Type> TypeAndGenericTypes(Type t) {
        var results = new List<Type>();

        results.Add(t);

        if (t.IsGeneric || t.IsEnumerable){
            foreach(var subtypes in t.TypeArguments){
                 results.AddRange(TypeAndGenericTypes(subtypes));
            }
        }
        return results;
    }
    IEnumerable<Type> Modules(Class c){
        return c.Methods
                .SelectMany(m => MethodDependencies(m, false))
                .GroupBy(G => G.OriginalName)
                .Select(group => group.First())
                .ToList();
    }

**Comment by DenisPitcher** [comment]
This issue may be worth checking out for the generation of dependencies of a specific controller's methods 
https://github.com/frhagn/Typewriter/issues/213

I needed to do something similar but to generate typescript including dependencies for a series of POCO contracts so I've modified @michaelaird's solution (https://gist.github.com/michaelaird/cfd15cebe92585556be2a14abff9dff4)

This is what I came up with which seems to fit my needs.  Of note, I needed to set a counter to limit the depth of recursion as I ran into issues without the limit so I arbitrarily set it to 10.
https://gist.github.com/DenisPitcher/4b895e3be5be9efa8f14529df1dbb1e4
